### DISCUSSION TITLE
Guarantee non-empty iterators

### DISCUSSION THREAD

**Discussion by thomasaarholt** [microsoft/pyright#5938] [Q&A] [answered]
Is there some way to assert that an iterator is non-empty? 

There is an excellent blog post called [Parse, don't validate](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/) that covers how in Haskell, the type signature `[a] -> a` isn't compilable, because it doesn't cover the case of an empty list, e.g. `[] -> a`. Instead you introduce a nonempty guarantee: `NonEmpty a -> a`, and it can compile.

It would be nice to be able to guarantee at least one entry, as is referred to in #2033. I see there is some [interesting discussion over at mypy](https://github.com/python/mypy/issues/9745).

**Comment by erictraut** [comment, marked as answer]
There's currently no mechanism in the Python type system that allows one to specify that an iterator must be non-empty. Such a mechanism would need to be specified in a PEP. The [python/typing discussion forum](https://github.com/python/typing/discussions) would be an appropriate place to start that discussion.

**Reply by thomasaarholt** [reply]
Thank you, that makes sense, I'll take my discussion there!
